### ISSUE TITLE
Imply generator send types from decorators

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#1656]
### Is your feature request related to a problem?

Probably a niche case, but consider this decorator:

```python
from functools import wraps
from typing import Any, Callable, Generator, TypeVar

TRet = TypeVar("TRet")

def commit(val: Any) -> str: ...

def storage(fn: Callable[..., Generator[Any, str, TRet]]) -> Callable[..., TRet]:
    @wraps(fn)
    def inner(*args, **kwargs) -> TRet:
        gen = fn(*args, **kwargs)
        try:
            val = next(gen)
            while True:
                name = commit(val)
                val = gen.send(name)
        except StopIteration as ex:
            return ex.value
    return inner
```

This consumes a generator, storing each yielded value under some key, and passes each key back to the decorated function.

A function making use of this generator might look something like this:

```python
def subtask1() -> int: ...
def subtask2(key: str) -> int: ...

@storage
def task():
    key = yield subtask1()
    yield subtask2(key)
    return bool(key)
```

Once `@storage` is applied, the type of `task()` becomes `() -> bool` (i.e. it's not seen as a generator from the outside).

Before it's decorated, the type of `task()` should be `Generator[int, str, bool]`.

In reality, `task()` is of type `Generator[int, None, bool]`, because there's no type information inside the function to say what the value of the yield expression is.  Pylance gives the following type warning:

```
Argument of type "() -> Generator[int, None, bool]" cannot be assigned to parameter "fn" of type "(*args: Any, **kwargs: Any) -> Generator[Any, str, TRet@storage]" in function "storage"
  Type "() -> Generator[int, None, bool]" cannot be assigned to type "(*args: Any, **kwargs: Any) -> Generator[Any, str, TRet@storage]"
    Function return type "Generator[int, None, bool]" is incompatible with type "Generator[Any, str, TRet@storage]"
      TypeVar "_T_contra@Generator" is contravariant
        Cannot assign to "None"
```

The type warning can be resolved by explicitly setting the type of `task()`:

```python
@storage
def task() -> Generator[int, str, bool]: ...
```

### Describe the solution you'd like

It would be nice if the send type of the underlying generator function could be implied by the decorator, given here it knows the wrapped function should have a `str` send type.

If this isn't feasible, perhaps the send type should at least be set to something other than `None` when the value of a yield expressions is used (or just always when not explicitly defined), be it `Unknown` or `Any`.

**Comment by erictraut** [comment]
Type inference doesn't work in that direction. You can't infer the type of an argument from a parameter type. In the Python type system, parameter types are specified, and argument types are verified against the known parameter types. You'll need to specify the generator type explicitly in this case.

```python
@storage
def task() -> Generator[int, str, bool]:
    key = yield subtask1()
    yield subtask2(key)
    return bool(key)
```
